Source/Core/InputCommon/ControllerInterface/evdev/evdev.cpp

```diff
@@ -207,30 +214,54 @@ evdevDevice::evdevDevice(const std::string& devnode) : m_devfile(devnode)
     {
       AddAnalogInputs(new Axis(num_axis, axis, false, m_dev),
                       new Axis(num_axis, axis, true, m_dev));
-      num_axis++;
+      ++num_axis;
     }
 
-  // Force feedback
+  // Disable autocenter
+  if (libevdev_has_event_code(m_dev, EV_FF, FF_AUTOCENTER))
+  {
+    input_event ie = {};
+    ie.type = EV_FF;
+    ie.code = FF_AUTOCENTER;
+    ie.value = 0;
+
+    static_cast<void>(write(m_fd, &ie, sizeof(ie)));
+  }
+
+  // Constant FF effect
+  if (libevdev_has_event_code(m_dev, EV_FF, FF_CONSTANT))
+  {
+    AddOutput(new ConstantEffect(m_fd));
+  }
+
+  // Periodic FF effects
   if (libevdev_has_event_code(m_dev, EV_FF, FF_PERIODIC))
   {
-    for (auto type : {FF_SINE, FF_SQUARE, FF_TRIANGLE, FF_SAW_UP, FF_SAW_DOWN})
-      if (libevdev_has_event_code(m_dev, EV_FF, type))
-        AddOutput(new ForceFeedback(type, m_dev));
+    for (auto wave : {FF_SINE, FF_SQUARE, FF_TRIANGLE, FF_SAW_UP, FF_SAW_DOWN})
+    {
+      if (libevdev_has_event_code(m_dev, EV_FF, wave))
+        AddOutput(new PeriodicEffect(m_fd, wave));
+    }
   }
+
+  // Rumble (i.e. Left/Right) (i.e. Strong/Weak) effect
   if (libevdev_has_event_code(m_dev, EV_FF, FF_RUMBLE))
   {
-    AddOutput(new ForceFeedback(FF_RUMBLE, m_dev));
+    // Strong motor:
+    AddOutput(new RumbleEffect(m_fd, true));
```

Comment:
What do you think of replacing the bool parameter with an enum class?

    AddOutput(new RumbleEffect(m_fd, RumbleStrength::Strong));
    AddOutput(new RumbleEffect(m_fd, RumbleStrength::Weak));

It would be clearer IMO (and removes the need for the comments). (If you do decide to change this, please do it in a separate commit… reviewing this would've been / would be easier with smaller commits.)

Reply:
Using an enum class did cross my mind. I'll change it.